src/VecSim/algorithms/hnsw/hnsw_tiered.h

```diff
@@ -803,14 +814,18 @@ int TieredHNSWIndex<DataType, DistType>::deleteVector(labelType label) {
 // 3. label exists in both indexes - we may have some of the vectors with the same label in the flat
 //    buffer only and some in the Main index only (and maybe temporal duplications).
 //    So, we get the distance from both indexes and return the minimum.
+
+// IMPORTANT: this should be called when the *tiered index locks are locked for shared ownership*,
+// along with HNSW index data guard lock. That is since the internal getDistanceFrom calls access
+// the indexes' data, and it is not safe to run insert/delete operation in parallel. Also, we avoid
+// acquiring the locks internally, since this is usually called for every vector individually, and
+// the overhead of acquiring and releasing the locks is significant in that case.
```

Comment:
I don't think that this is the (only) location for this comment. Should be part of the API.